lib/CodeGen/CodeGenFunction.h

```diff
@@ -1238,6 +1238,48 @@ class CodeGenFunction : public CodeGenTypeCache {
   const TargetInfo &getTarget() const { return Target; }
   llvm::LLVMContext &getLLVMContext() { return CGM.getLLVMContext(); }
 
+public:
+  enum BoundsCheckKind {
+    BC_None,
+    BC_Deref,
+    BC_ArraySubscript
+  };
+
+private:
+  BoundsExpr* NextBoundsExprForCheck;
+  BoundsCheckKind NextBoundsCheckKind;
+
+public:
+  BoundsExpr* GetAndClearNextBoundsCheckExpr(BoundsCheckKind Kind) {
+    if (!getLangOpts().CheckedC)
+      return nullptr;
+
+    if (NextBoundsCheckKind == BC_None)
+      return nullptr;
+
+    if (NextBoundsCheckKind != Kind)
+      return nullptr;
+
+    BoundsExpr* result = NextBoundsExprForCheck;
+    NextBoundsExprForCheck = nullptr;
+    NextBoundsCheckKind = BC_None;
+
+    return result;
+  }
+
+  void SetNextBoundsCheckExpr(BoundsExpr *Bounds, BoundsCheckKind Kind) {
+    if (!getLangOpts().CheckedC)
+      return;
+
+    assert(!Bounds->isInvalid() && "Only valid bounds checks can be checked against");
```

Comment:
You need to handle the case where Bounds is null here - right now if Bounds is null, you'll try to call a method on it.